### DISCUSSION TITLE
Tract envelope

### DISCUSSION THREAD

**Discussion by peterneher** [dipy/dipy#2552] [Tractography / Tractometry]
Hi everyone,

Is it possible to calculate a binary tract envelope using dipy? I hope I did not miss something obvious in the documentation :)

Cheers,
Peter

**Comment by arokem** [comment]
I don't think that we have functionality already implemented for this, but it shouldn't be too hard to create this from what's already here. For clarification: what does an "envelope" mean in this case? A binary mask with ones wherever there is a tract coordinate? Or the convex hull of the coordinates of a tract?

**Reply by peterneher** [reply]
Yes, simply a binary mask of the tract. Of course you are right that it would be straight-forward to simply set every voxel to 1 that contains a streamline point, but this does not cover the complete tract since voxels can be traversed by a streamline without containig a streamline point. I implemented this in MITK Diffusion but now I would like to use it in dipy. If I get around to it, maybe I will contribute this.

**Reply by arokem** [reply]
Makes sense! Can you point to the MITK Diffusion implementation? Looks like MITK Diffusion is under BSD, so it shouldn't be too hard to create an implementation based on what you did there. Thanks!

**Comment by Garyfallidis** [comment]
This seems like something that we already have @peterneher and @arokem. You have to look at JC's implementation of track counts. One sec ... digging ...

**Reply by peterneher** [reply]
Let me know if you find something. Otherwise, I already implemented it, so I could contribute it if you like.

@arokem , here is the MITK implementation of the tract density image filter that also calculates the envelope: https://github.com/MIC-DKFZ/MITK-Diffusion/blob/master/Modules/FiberTracking/Algorithms/itkTractDensityImageFilter.cpp. This filter intersects each line segment of each streamline with the voxels inside of the bounding-box of the respective segment (using vtk). Here is the position where this happens: https://github.com/MIC-DKFZ/MITK-Diffusion/blob/master/Modules/FiberTracking/Algorithms/itkTractDensityImageFilter.cpp#L142 

I implemented the python version analogously, also using vtk.

**Comment by Garyfallidis** [comment]
Apologies, busy days yesterday and unfortunately today too. I believe @jchoude implemented a similar function a few years back that was already cythonized. @jchoude can you remind us if your implementation is what we are looking for?

**Reply by Garyfallidis** [reply]
@peterneher, I talked today with @jchoude and he will get back to you in a few days after completing a deadline. If there is still no solution then your contribution would be much appreciated.

**Comment by jchoude** [comment]
Hi @peterneher (long time no talk!) and @Garyfallidis ,

I double checked. The underlying algorithm is implemented in https://github.com/dipy/dipy/blob/master/dipy/tracking/vox2track.pyx#L229
but is formulated to be used in a mask intersection scenario.

I had an extremely old branch that generalized this to return the density map (robust tract counts) instead, which could then be binarized as the tract envelope.

The cleanest version now exists in Scilpy: https://github.com/scilus/scilpy/blob/master/scilpy/tractanalysis/streamlines_metrics.pyx#L39

This could probably be integrated to Dipy. On my side, I sadly don't have time to code now, but the Scilpy version is just a cleaned up version of my old code.

Hope that helps!
